### ISSUE TITLE
HowTo: Nested Dictionaries

### ISSUE THREAD

**Issue by ShurayukiHime** [pikepdf/pikepdf#312]
Hi,
Thank you for developing pikepdf.
I could not find in the documentation any explanation on how to embed Dictionaries within each other, as well as other elements (e.g., Rectangles). A simple example of what I want to achieve is the following:
```
<</Type /Annot 
/Subtype /Link 
/Rect [184.25 479.06 405.35 422.36] 
/A <<
/Type Action
/S /URI
/URI (google.com)
>>
>>
```
Which is, in turn, referenced by another object as, e.g., a Page.

I found in the docs the API `make_indirect`, but I am not sure if I it is the one to use in this case, and how.

Thanks.

**Comment by jbarlow83** [comment]
Use Object.parse if you want to construct objects from Postscript object notation. 

Also

```
pikepdf.Dictionary(Nested=pikepdf.Dictionary(Key1=...))

```

will create nested direct object dictionaries. Use `make_indirect` when the PDF spec requires an indirect object reference, typically when a pointer to an existing object is required rather than a copy.

**Comment by ShurayukiHime** [comment]
Thanks!
Supposing I have this object
```
<<
/Type /StructElem
/Pg 94 0 R
/K 75 0 R
>>
```
where 94 is the reference to the Page object, and 75 to an OBJR object (needed when I declare a Dictionary within an object stream), e.g.,
```
<< /Obj 77 0 R
/Pg 94 0 R
/Type /OBJR >>
```
- How can I use `make_indirect` to obtain the references of the Page and OBJR objects?
- Is there a way to handle references of Dictionaries within object streams? In this case, object 77.

Thanks again for your help!

**Comment by jbarlow83** [comment]
1. `make_indirect` converts a direct object into an indirect object. If already an indirect object, it turns the object without modification.

Use `object.objgen` to obtain the object, generation pair if needed (e.g. 77, 0) and `Pdf.get_object` to look up an object in this way. e.g. if `structelem` is assigned to your object, `structelem['/Pg']` returns the object (94, 0), as would `pdf.get_object(94, 0)`.

2. pikepdf processes object streams as if it decompressed them all into the objects they describe. Object streams are transparently loaded, and generated automatically when a PDF is saved if appropriate. Essentially you don't have to worry about object streams at all, pikepdf takes care of them for you.

Stream objects are not object streams - the terminology chosen by the PDF specification is unfortunate. Stream objects must be indirect objects wherever they are referred to. pikepdf does not allow you to create a direct stream object. Stream objects can only be created indirect.

**Comment by ShurayukiHime** [comment]
Hi,
thanks again for the very helpful answer. I think I managed to create the support structure for the NameTree, as well as StructElement and StructTreeRoot.

I have one last question:
The object in the stream needs an object reference, which is going to be mentioned in the StructElement (e.g., see above, `77 0 R`).
**Is there a way to get an available / correct object reference via pikepdf, without actually attaching the object to the pdf?**

I tried creating an object, making it indirect, getting its reference and then deleting it via the `del` statement. Unfortunately, the reference number collided with one of an already existing object (I'm using pikepdf to modify an existing PDF).

Thank you.
